integrationtests/server_test.go

```diff
@@ -0,0 +1,201 @@
+		Expect(err).ToNot(HaveOccurred())
+		_, err = f.Write([]byte("X-Original-Url: https://quic.clemente.io:" + port + "/" + filename + "\n"))
+		Expect(err).ToNot(HaveOccurred())
+		_, err = f.Write([]byte("Content-Length: " + strconv.Itoa(len(data)) + "\n\n"))
+		Expect(err).ToNot(HaveOccurred())
+		_, err = f.Write(data)
+		Expect(err).ToNot(HaveOccurred())
+	}
+
+	// download files must be create *before* the quic_server is started
+	// the quic_server reads its data dir on startup, and only serves those files that were already present then
+	startServer := func() {
+		go func() {
+			defer GinkgoRecover()
+			var err error
+			command := exec.Command(
+				serverPath,
+				"--quic_response_cache_dir="+filepath.Join(tmpDir, "quic.clemente.io"),
+				"--key_file="+filepath.Join(tmpDir, "key.pkcs8"),
+				"--certificate_file="+filepath.Join(tmpDir, "cert.pem"),
+				"--port="+port,
+			)
+			session, err = Start(command, GinkgoWriter, GinkgoWriter)
+			Expect(err).NotTo(HaveOccurred())
+		}()
+	}
+
+	stopServer := func() {
+		session.Kill()
+	}
+
+	BeforeEach(func() {
+		port = strconv.Itoa(20000 + int(mrand.Int31n(10000)))
+
+		var err error
+		tmpDir, err = ioutil.TempDir("", "quic-server-certs")
+		Expect(err).ToNot(HaveOccurred())
+
+		// generate an RSA key pair for the server
+		key, err := rsa.GenerateKey(rand.Reader, 1024)
+		Expect(err).ToNot(HaveOccurred())
+		// save the private key in PKCS8 format to disk (required by quic_server)
+		pkcs8key, err := marshalPKCS8PrivateKey(key)
+		Expect(err).ToNot(HaveOccurred())
+		f, err := os.Create(filepath.Join(tmpDir, "key.pkcs8"))
+		Expect(err).ToNot(HaveOccurred())
+		_, err = f.Write(pkcs8key)
+		Expect(err).ToNot(HaveOccurred())
+		f.Close()
+
+		// generate a Certificate Authority
+		// this CA is used to sign the server's key
+		// it is set as a valid CA in the QUIC client
+		rootKey, CACert := generateCA()
+		template := &x509.Certificate{
+			SerialNumber: big.NewInt(1),
+			NotBefore:    time.Now(),
+			NotAfter:     time.Now().Add(time.Hour),
+			Subject:      pkix.Name{CommonName: "quic.clemente.io"},
+		}
+		certDER, err := x509.CreateCertificate(rand.Reader, template, CACert, &key.PublicKey, rootKey)
+		Expect(err).ToNot(HaveOccurred())
+		// save the certificate to disk
+		certOut, err := os.Create(filepath.Join(tmpDir, "cert.pem"))
+		Expect(err).ToNot(HaveOccurred())
+		pem.Encode(certOut, &pem.Block{Type: "CERTIFICATE", Bytes: certDER})
+		certOut.Close()
+
+		// prepare the h2quic.client
+		certPool := x509.NewCertPool()
+		certPool.AddCert(CACert)
+		client = &http.Client{
+			Transport: &h2quic.QuicRoundTripper{
+				TLSClientConfig: &tls.Config{RootCAs: certPool},
+			},
+		}
+	})
+
+	AfterEach(func() {
+		Expect(tmpDir).ToNot(BeEmpty())
+		err := os.RemoveAll(tmpDir)
+		Expect(err).ToNot(HaveOccurred())
+		tmpDir = ""
+	})
+
+	It("downloads a hello", func(done Done) {
+		data := []byte("Hello world!\n")
+		createDownloadFile("hello", data)
+
+		startServer()
```

Comment:
Should we move this to Before / AfterEach? We could then get rid of the startServer / stopServer functions.